example_test.go

```diff
@@ -10,6 +10,131 @@ import (
 	"inet.af/netaddr"
 )
 
+func ExampleIP() {
+	ip, err := netaddr.ParseIP("192.0.2.3")
+	if err != nil {
+		panic(err)
+	}
+
+	// netaddr.IP supports comparison using ==
+	fmt.Println(ip == netaddr.IPv4(192, 0, 2, 3))
+
+	// netaddr.IP can be used as a map key
+	hosts := make(map[netaddr.IP]string)
+	hosts[ip] = "example.net"
+	fmt.Println(hosts)
+	// Output:
+	// true
+	// map[192.0.2.3:example.net]
+}
+
+func ExampleIP_Is4() {
+	ipv4 := netaddr.MustParseIP("192.0.2.3")
+	ipv6 := netaddr.MustParseIP("2001:db8::68")
+	ipEmpty := netaddr.IP{}
+	ipv4MappedToIPv6 := netaddr.MustParseIP("::ffff:192.0.2.3")
+
+	fmt.Printf("ipv4.Is4() -> %v\n", ipv4.Is4())
+	fmt.Printf("ipv6.Is4() -> %v\n", ipv6.Is4())
+	fmt.Printf("ipEmpty.Is4() -> %v\n", ipEmpty.Is4())
+	fmt.Printf("ipv4MappedToIPv6.Is4() -> %v\n", ipv4MappedToIPv6.Is4())
+	// Output:
+	// ipv4.Is4() -> true
+	// ipv6.Is4() -> false
+	// ipEmpty.Is4() -> false
+	// ipv4MappedToIPv6.Is4() -> false
+}
+func ExampleIP_Is4in6() {
+	ipv4 := netaddr.MustParseIP("192.0.2.3")
+	ipv6 := netaddr.MustParseIP("2001:db8::68")
+	ipEmpty := netaddr.IP{}
+	ipv4MappedToIPv6 := netaddr.MustParseIP("::ffff:192.0.2.3")
+
+	fmt.Printf("ipv4.Is4in6() -> %v\n", ipv4.Is4in6())
+	fmt.Printf("ipv6.Is4in6() -> %v\n", ipv6.Is4in6())
+	fmt.Printf("ipEmpty.Is4in6() -> %v\n", ipEmpty.Is4in6())
+	fmt.Printf("ipv4MappedToIPv6.Is4in6() -> %v\n", ipv4MappedToIPv6.Is4in6())
+	// Output:
+	// ipv4.Is4in6() -> false
+	// ipv6.Is4in6() -> false
+	// ipEmpty.Is4in6() -> false
+	// ipv4MappedToIPv6.Is4in6() -> true
+}
+func ExampleIP_Is6() {
```

Comment:
There's a lot of duplication between these. I wonder whether they'd be easier to follow as a part of ExampleIP. Something like:

```go
  examples := []struct{
  name string
  ip IP
  }{
    {name: "ipv4", ip: netaddr.MustParseIP("192.0.2.3")},
    // and so on
  }

  for _, example := range examples {
    fmt.Printf("%s: String=%v Is4=%v Is6=%v Is4in6=%v Is Zero=%v\n", example.name, example.ip.String(), /* ... */)
  }
}
```

Maybe go really wild and use a text.tabwriter to make a pretty table. This would be less useful for copy/paste, but easier to see the big picture of what an IP is and what all of its methods mean without having to hop around between examples.

What do you think?

What do you think, Brad/Dave/Matt?

Reply:
I think it's useful to have an example against each method, because people reading the docs are much more likely to find the example for method M in situ - next to the description.

I like the idea of a table in the `IP` example, in addition to the individual method examples. It's a slightly higher maintenance burden, for a much higher end user benefit.

Reply:
I've kept the individual method examples, and added a table to `ExampleIP`

Reply:
I like the table example.